src/main/scala/zio/prelude/Idempotent.scala

```diff
@@ -0,0 +1,498 @@
+package zio.prelude
+
+import zio.prelude.coherent.EqualIdempotent
+import zio.test.TestResult
+import zio.test.laws.{ Lawful, Laws }
+
+/**
+ * The `Idempotent` type class describes a binary operator for a type `A` that
+ * is both associative and produces the same value when combining two identical values.
+ * This means that `a <> a` is equal to `a` for all values `a`.
+ * Example of idempotent operations is union of sets, but not addition of integers.
+ *
+ * Idempotent operators are useful because combining  the values with an idempotent
+ * operation results in the same value regardless of the number of values
+ * are combined, allowing us to optimize out unnecessary combinations of the same values.
+ */
+trait Idempotent[A] extends Associative[A] { self =>
+
+  /**
+   * If there are many Equal elements in  sequence, it will remove them all except for the first one.
+   * `a a b c c c a` will result in `a b c a`.
+   */
+  def optimize[T[+_]: Traversable](elems: T[A])(implicit A: Equal[A]): List[A] = {
```

Comment:
I wonder if this method really belongs here. It requires `Equal` and `Traversable` but doesn't actually use the functionality of `Associative` or `Idempotent` at all. I think this might make more since as a method on `Traversable`, `distinct` and then the fact that something is `Idempotent` just tells us that calling distinct on it and then reducing it with an idempotent operation is identical to reducing the original collection, just like we can fold left or fold right with an associative operation.

Reply:
Good point. Moved.
I named it `dropRedundant` though, because `distinct` has the connotation that it creates as-if a Set, but that's not the case here.

Reply:
Sorry can you say more about what you mean? `distinct` normally returns the same collection type so like called `distinct` on a `List` would return a `List` and maintain the ordering. We can't quite honor the returns the same collection type guarantee because `Traversable` only allows us to rebuild a structure with the shape as the original structure but I think returning a `List` or `Chunk` is the closest thing.

Reply:
`List(1,2,2,1).distinct` == `List(1,2)`
`List(1,2,2,1).dropRedundant` == `List(1,2,1)`

Reply:
Right. Since it isn't commutative we can't rearrange the terms to further simplify. Goes back to that question of whether it should extend commutative or not. If it was commutative then we could remove all the duplicates.